tapa.py

```diff
@@ -0,0 +1,163 @@
+        else:
+            return False
+
+    return True
+
+
+def iterate_bitmasks_for_clues(clues, num_bits):
+    """
+    Returns an iterator yielding the subset of integers in [0, 2^num_bits)
+    which, when interpreted as a bitmask, would be valid as the fill state of
+    cells surrounding a cell with the given clues.
+    """
+    bound = 1 << num_bits
+    high_bit = bound >> 1
+    mask = bound - 1
+    for bits in range(bound):
+        orig_bits = bits
+        groups = []
+        if bits == mask:
+            groups.append(num_bits)
+        elif bits != 0:
+            # Rotate the bits until the high bit is clear so we don't have to
+            # worry about bit groups wrapping around (we are guaranteed that at
+            # least one bit is clear because we just handled the bits == mask
+            # case)
+            while bits & high_bit != 0:
+                bits = (bits << 1) & mask | 1
+
+            count = 0
+            while bits > 0:
+                if bits & 1 == 1:
+                    count += 1
+                else:
+                    if count > 0:
+                        groups.append(count)
+                    count = 0
+                bits >>= 1
+            if count > 0:
+                groups.append(count)
+        if do_groups_match_clues(groups, clues):
+            yield orig_bits
+
+
+def solve_tapa(puzzle):
+    lines = [
+        [
+            [
+                char for char in cell if char != ' '
+            ] for cell in line.split('|')
+        ] for line in puzzle.strip().split('\n') if line != ''
+    ]
+
+    # Each cell in this grid will have an integer variable which is >= 0 if the
+    # cell is filled and < 0 otherwise. (The specific values within those
+    # ranges are only relevant for the one-contiguous-region constraints.)
+    g = Grid(len(lines[0]), len(lines))
+
+    def neighbor_is_filled(c):
+        return not isinstance(c, Invalid) and c.var >= 0
+
+    s = Solver()
+
+    for cell in g.cells:
+        clues = lines[cell.y][cell.x]
+        if clues:
+            # Clued cells may not be filled
+            s.add(cell.var < 0)
+
+            if clues != ['*']:
+                # Constrain the surroundings of a cell by its clues
+                surrounding_cells = get_surrounding_cells(cell)
+                num_bits = len(surrounding_cells)
+                s.add(Or([
+                    And([
+                        neighbor_is_filled(cell2) == (bits & (1 << i) != 0)
+                        for i, cell2 in enumerate(surrounding_cells)
+                    ]) for bits in iterate_bitmasks_for_clues(clues, num_bits)
+                ]))
+        else:
+            # Filled cells must form one contiguous region (part 1)
+            s.add(Or(cell.var <= 0, *(
+                And(n.var >= 0, cell.var > n.var) for n in cell.neighbors()
+            )))
+
+    # Filled cells must form one contiguous region (part 2)
+    s.add(Distinct([c.var for c in g.cells]))
```

Comment:
This is fine and it works, but I'm curious if it's more or less efficient to relax this constraint to be "at most one 0", and in return tighten the constraint at L120 to be "positive cells must have a non-negative neighbor exactly 1 lower", which is roughly how I did it in Galaxies. Actually I guess you wouldn't even need to tighten up L120; "at most one 0" is actually all you need out of this `Distinct()`.

Reply:
But I think tightening L120 will help Z3 by giving it fewer options, probably. Who knows. Z3 performance is completely mysterious.

Reply:
I just tested and couldn't measure any performance difference doing it one way or the other, but I do know that the `AtLeast`/`AtMost` operators cause Z3 to use a more complex solver, while `Distinct` constraints can be solved in the core SAT solver, so I tend to reach for `Distinct` first. That's pretty much all I know on the matter.

Reply:
Huh. I didn't know that about a different solver, although it makes sense. Good to know. I assume `PbEq` and friends similarly require the advanced solver? (I think `AtMost` is secretly implemented in terms of `PbLt` anyway.)

Reply:
Yeah, I don't know about one being implemented in terms of the other, but my understanding is that they all require the same pseudo-Boolean machinery.

Although that machinery might just be the right thing to use. Last night I only compared the existing code to using `AtMost` and tightening L120, but I just now tried the third possibility of using `AtMost` but leaving L120 alone, and *that* did shave off about 25% of the in-Z3 time. (Which is, admittedly, a fraction of a second—but it's the principle of the thing I guess? It's a good vindication of the `AtMost` approach at least.)